File: Docusnap/database/modules/test_taught_ref_disagree_suppress.js
'use strict';
/*
 * test_taught_ref_disagree_suppress.js — the trust.js SEAM pins for TAUGHT_REF_DISAGREE_SUPPRESS
 * (mig 186, 2026-09-19; gary → Oracle SIGN-OFF-W/COND). Engine-side the filter moves an off-shape
 * ref competitor out of `disagree`/`discounted` into a NEW `suppressed_taught_role` key. These pins
 * lock the TWO trust.js contracts that make that safe:
 *   1. THE HOLD LIFTS: _pageFamilyDisagrees does NOT scan `suppressed_taught_role` (so the
 *      role-disagreement refusal no longer fires), but STILL fires on a real `disagree`.
 *   2. THE LICENCE DOES NOT LAUNDER (Oracle C1, the ship-blocker): _corrobLicensed returns FALSE on
 *      any non-empty `suppressed_taught_role` — so emptying `disagree` can never flip a mapping+crop
 *      common-mode record into a silent corroborated auto-file. refBadgeVerified inherits it.
 *
 * Run:  node database/modules/test_taught_ref_disagree_suppress.js
 */
const path = require('path');
const trust = require(path.join(__dirname, 'trust.js'));

let P = 0, F = 0;
const ok = (name, cond) => { if (cond) { P++; console.log('  ok  ' + name); } else { F++; console.log('  FAIL ' + name); } };

console.log('1. _corrobLicensed — the suppressed-entry guard (Oracle C1, the ship-blocker)');
{
  // The exact bug shape: a taught MAPPING winner with an agreeing CROP candidate (common-mode, Pelican-565),
  // and the only thing that had been holding the licence off was the now-emptied `disagree`.
  const laundered = { winner_family: 'mapping', agree: ['crop'], disagree: [],
                      suppressed_taught_role: [{ family: 'keyword', value: 'JB-2554' }], independent_agree: true };
  ok('a record whose competitor was moved to suppressed_taught_role is NOT corrob-licensed (no silent auto-file)',
     trust._corrobLicensed(laundered) === false);
  // CONTROL: the SAME record without the suppressed entry IS licensed — proves the guard is the deciding line,
  // not some other predicate (so a future dev who removes the guard sees this flip and the pin above go red).
  const control = { winner_family: 'mapping', agree: ['crop'], disagree: [], independent_agree: true };
  ok('CONTROL: the same mapping+crop record with NO suppressed entry IS licensed', trust._corrobLicensed(control) === true);
  // A stringified record (the stored column form) is guarded too.
  ok('guard applies to the JSON-string record form', trust._corrobLicensed(JSON.stringify(laundered)) === false);
  // An empty suppressed list must NOT block (byte-identical to no key).
  ok('an EMPTY suppressed_taught_role list does not block the licence',
     trust._corrobLicensed({ winner_family: 'mapping', agree: ['crop'], disagree: [], suppressed_taught_role: [], independent_agree: true }) === true);
}

console.log('\n2. _pageFamilyDisagrees — the hold LIFTS for a suppressed competitor, but a real disagree still holds');
{
  // Competitor ONLY in suppressed_taught_role (engine moved it there) → no page-family disagreement → hold lifts.
  const suppressed = { winner_family: 'mapping', agree: [], disagree: [],
                       suppressed_taught_role: [{ family: 'keyword', value: 'JB-2554' }], independent_agree: false };
  ok('a competitor in suppressed_taught_role is NOT seen by _pageFamilyDisagrees (the role-refusal stops holding)',
     trust._pageFamilyDisagrees(suppressed) === null);
  // A genuine disagree is unchanged — the guard did not weaken the refusal for the normal case.
  const held = { winner_family: 'mapping', agree: [], disagree: [{ family: 'keyword', value: 'JB-2554' }], independent_agree: false };
  const hit = trust._pageFamilyDisagrees(held);
  ok('a real disagree still fires _pageFamilyDisagrees (normal hold intact)', hit && hit.family === 'keyword' && hit.value === 'JB-2554');
  // And a competitor still in `discounted` still holds (only the engine, when it actually suppresses, moves it out).
  const disc = { winner_family: 'mapping', agree: [], disagree: [], discounted: [{ family: 'keyword', value: 'JB-2554', reason: 'x' }], independent_agree: false };
  ok('a competitor still in discounted still fires _pageFamilyDisagrees', trust._pageFamilyDisagrees(disc) && trust._pageFamilyDisagrees(disc).family === 'keyword');
}

console.log('\n3. refBadgeVerified — inherits the guard (no green badge via the corroborated leg on a suppressed record)');
{
  const doc = { supplier_name: 'Castellan Security Systems', type_slug: 'service_worksheet' };
  const extLaundered = { extraction_method: 'template_mapping', corroboration: JSON.stringify(
    { winner_family: 'mapping', agree: ['crop'], disagree: [], suppressed_taught_role: [{ family: 'keyword', value: 'JB-2554' }], independent_agree: true }) };
  // No DB history passed (opts.formats=[]) and method is not an authoritative literal, so the ONLY road to
  // "verified" would be the corroborated leg — which the guard closes. Verdict must be false.
  ok('refBadgeVerified is NOT true via corroboration for a suppressed-competitor record',
     trust.refBadgeVerified(null, doc, 'reference_number', extLaundered, { formats: [] }) === false);
  // INVARIANCE: the control (no suppressed entry) WOULD be corroborated-verified — proves the split rides the guard.
  const extControl = { extraction_method: 'template_mapping', corroboration: JSON.stringify(
    { winner_family: 'mapping', agree: ['crop'], disagree: [], independent_agree: true }) };
  ok('INVARIANCE: the same record without the suppressed entry IS corroborated-verified',
     trust.refBadgeVerified(null, doc, 'reference_number', extControl, { formats: [] }) === true);
}

console.log('\n' + (F === 0 ? 'ALL PASS' : F + ' FAILED') + `  (${P} ok)`);
process.exit(F ? 1 : 0);

File: Docusnap/python_backend/tests/test_taught_ref_disagree_suppress.py
"""
test_taught_ref_disagree_suppress.py — the pure record-transform pins for TAUGHT_REF_DISAGREE_SUPPRESS
(mig 186, 2026-09-19; gary -> Oracle SIGN-OFF-W/COND). Tests engine._suppress_taught_ref_disagree_record
directly: it moves an OFF-learned-shape ref competitor out of disagree/discounted into a NEW
`suppressed_taught_role` key, but ONLY when the winner is an authoritative taught read that itself
matches the learned shape. off_shape() is injected, so these pins lock the MOVE LOGIC and the
accepted-trade-off holds (garbled winner / same-shape competitor / non-taught winner => NO move => the
doc still HOLDS, fail-toward-review). The real off_shape / ref-only scoping / date-exclusion live at the
engine call site and are exercised by the realdoc harness.

Run:  py -3.12 python_backend/tests/test_taught_ref_disagree_suppress.py
"""
import os
import sys

try:
    sys.stdout.reconfigure(encoding='utf-8')
except Exception:
    pass
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from extraction.engine import _suppress_taught_ref_disagree_record, _is_stage05_located  # noqa: E402

_fail = 0


def ok(name, cond):
    global _fail
    print(("  ok  " if cond else "  FAIL ") + name)
    if not cond:
        _fail += 1


# A stand-in for "value is off the scope's learned shape": the learned shape is CJB-#### (letters+digits),
# so the taught winner CJB-1578 is ON-shape and the "Job Ref" competitor JB-2554 is OFF-shape (shorter
# letter run). This mirrors the real format_anomaly_checker shape test the call site injects, and pins the
# CJB-vs-JB letter-length discrimination the OCR note calls out.
def off_shape_cjb(v):
    v = str(v or "").strip()
    if not v:
        return False
    return not v.startswith("CJB-")   # JB-2554 -> True (off shape); CJB-1600 -> False (same shape)


print("1. TAUGHT WINS: an off-shape keyword competitor is moved out of disagree into suppressed_taught_role")
rec = {"winner_family": "mapping", "agree": [], "disagree": [{"family": "keyword", "value": "JB-2554"}],
       "independent_agree": False}
moved = _suppress_taught_ref_disagree_record(rec, True, True, off_shape_cjb)
ok("one entry moved", moved == 1)
ok("disagree emptied", rec["disagree"] == [])
ok("suppressed_taught_role carries the competitor", rec.get("suppressed_taught_role") == [{"family": "keyword", "value": "JB-2554"}])

print("\n2. SAFETY — garbled taught box (winner_shape_ok False): NO move, the doc still HOLDS")
rec = {"winner_family": "mapping", "agree": [], "disagree": [{"family": "keyword", "value": "JB-2554"}], "independent_agree": False}
moved = _suppress_taught_ref_disagree_record(rec, True, False, off_shape_cjb)
ok("nothing moved", moved == 0)
ok("disagree untouched (still holds)", rec["disagree"] == [{"family": "keyword", "value": "JB-2554"}])
ok("no suppressed key added", "suppressed_taught_role" not in rec)

print("\n3. SAFETY — non-taught winner (winner_taught False): NO move")
rec = {"winner_family": "keyword", "agree": [], "disagree": [{"family": "keyword", "value": "JB-2554"}], "independent_agree": False}
ok("nothing moved", _suppress_taught_ref_disagree_record(rec, False, True, off_shape_cjb) == 0)
ok("disagree untouched", rec["disagree"] == [{"family": "keyword", "value": "JB-2554"}])

print("\n4. SAFETY — SAME-shape competitor (genuine same-field ambiguity): NOT moved, still holds")
rec = {"winner_family": "mapping", "agree": [], "disagree": [{"family": "keyword", "value": "CJB-1600"}], "independent_agree": False}
moved = _suppress_taught_ref_disagree_record(rec, True, True, off_shape_cjb)
ok("nothing moved (CJB-1600 is same shape as the taught CJB-1578)", moved == 0)
ok("same-shape rival stays in disagree (still holds)", rec["disagree"] == [{"family": "keyword", "value": "CJB-1600"}])

print("\n5. Oracle C2 — a competitor already in `discounted` is moved too (else the hold stays live)")
rec = {"winner_family": "mapping", "agree": [], "disagree": [],
       "discounted": [{"family": "keyword", "value": "JB-2554", "reason": "alphanumeric_format_invalid"}], "independent_agree": False}
moved = _suppress_taught_ref_disagree_record(rec, True, True, off_shape_cjb)
ok("one entry moved from discounted", moved == 1)
ok("discounted emptied", rec["discounted"] == [])
ok("moved entry keeps its reason in suppressed_taught_role",
   rec.get("suppressed_taught_role") == [{"family": "keyword", "value": "JB-2554", "reason": "alphanumeric_format_invalid"}])

print("\n6. MIXED — only the off-shape competitor moves; a same-shape rival is left holding")
rec = {"winner_family": "mapping", "agree": [],
       "disagree": [{"family": "keyword", "value": "JB-2554"}, {"family": "crop", "value": "CJB-1600"}], "independent_agree": False}
moved = _suppress_taught_ref_disagree_record(rec, True, True, off_shape_cjb)
ok("only the off-shape entry moved", moved == 1)
ok("the same-shape crop rival stays in disagree (still holds)", rec["disagree"] == [{"family": "crop", "value": "CJB-1600"}])
ok("suppressed carries only the off-shape competitor", rec.get("suppressed_taught_role") == [{"family": "keyword", "value": "JB-2554"}])

print("\n7. NO-OP — an empty/agreeing record is byte-identical (returns 0, adds no key)")
rec = {"winner_family": "mapping", "agree": ["crop"], "disagree": [], "independent_agree": True}
before = dict(rec)
ok("nothing moved", _suppress_taught_ref_disagree_record(rec, True, True, off_shape_cjb) == 0)
ok("record unchanged", rec == before)

print("\n8. _is_stage05_located gate — authoritative taught methods only")
ok("template_mapping is Stage-0.5 located", _is_stage05_located("template_mapping") is True)
ok("template_mapping_inline (suffix) is located", _is_stage05_located("template_mapping_inline") is True)
ok("template_registration_expanded is located", _is_stage05_located("template_registration_expanded") is True)
ok("keyword is NOT located", _is_stage05_located("keyword") is False)
ok("template_fixed (auto-learned) is NOT located", _is_stage05_located("template_fixed") is False)
ok("None is NOT located", _is_stage05_located(None) is False)

print("\n" + ("ALL PASS" if _fail == 0 else f"{_fail} FAILED"))
sys.exit(1 if _fail else 0)
